drivers/i2c/i2c_rcar.c

```diff
@@ -0,0 +1,385 @@
+
+	/* Set slave address & transfer mode */
+	i2c_rcar_write(config, RCAR_I2C_ICMAR, (chip << 1) | read);
+	/* Reset */
+	i2c_rcar_write(config, RCAR_I2C_ICMCR, RCAR_I2C_ICMCR_MASTER | RCAR_I2C_ICMCR_ESG);
+	/* Clear Status */
+	i2c_rcar_write(config, RCAR_I2C_ICMSR, 0);
+
+	/* Wait for address & transfer mode to be transmitted */
+	if (read != 0) {
+		return i2c_rcar_wait_for_state(dev, RCAR_I2C_MAT | RCAR_I2C_MDR);
+	} else {
+		return i2c_rcar_wait_for_state(dev, RCAR_I2C_MAT | RCAR_I2C_MDE);
+	}
+}
+
+static int i2c_rcar_transfer_msg(const struct device *dev, struct i2c_msg *msg)
+{
+	const struct i2c_rcar_cfg *config = DEV_I2C_CFG(dev);
+	uint32_t i, reg;
+	int ret;
+
+	if ((msg->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ) {
+		/* Reading as master */
+		i2c_rcar_write(config, RCAR_I2C_ICMCR, RCAR_I2C_ICMCR_MASTER);
+
+		for (i = 0; i < msg->len; i++) {
+			if (msg->len - 1 == i) {
+				i2c_rcar_write(config, RCAR_I2C_ICMCR, RCAR_I2C_ICMCR_MASTER |
+					       RCAR_I2C_ICMCR_FSB);
+			}
+
+			/* Start data reception */
+			reg = i2c_rcar_read(config, RCAR_I2C_ICMSR);
+			reg &= ~RCAR_I2C_MDR;
+			i2c_rcar_write(config, RCAR_I2C_ICMSR, reg);
+
+			/* Wait for data to be received */
+			ret = i2c_rcar_wait_for_state(dev, RCAR_I2C_MDR);
+			if (ret != 0) {
+				return ret;
+			}
+
+			msg->buf[i] = i2c_rcar_read(config, RCAR_I2C_ICRXD_ICTXD) & 0xff;
+		}
+	} else {
+		/* Writing as master */
+		for (i = 0; i < msg->len; i++) {
+			i2c_rcar_write(config, RCAR_I2C_ICRXD_ICTXD, msg->buf[i]);
+
+			i2c_rcar_write(config, RCAR_I2C_ICMCR, RCAR_I2C_ICMCR_MASTER);
+
+			/* Start data transmission */
+			reg = i2c_rcar_read(config, RCAR_I2C_ICMSR);
+			reg &= ~RCAR_I2C_MDE;
+			i2c_rcar_write(config, RCAR_I2C_ICMSR, reg);
+
+			/* Wait for all data to be transmitted */
+			ret = i2c_rcar_wait_for_state(dev, RCAR_I2C_MDE);
+			if (ret != 0) {
+				return ret;
+			}
+		}
+	}
+
+	/* Finish the transfer */
+	if ((msg->flags & I2C_MSG_STOP) == I2C_MSG_STOP) {
+		ret = i2c_rcar_finish(dev);
+	}
+
+	return ret;
+}
+
+static int i2c_rcar_transfer(const struct device *dev,
+			     struct i2c_msg *msgs, uint8_t num_msgs,
+			     uint16_t addr)
+{
+	const struct i2c_rcar_cfg *config = DEV_I2C_CFG(dev);
+	uint16_t timeout = 0;
+	int ret;
+
+	if (!num_msgs) {
+		return 0;
+	}
+
+	/* Wait for the bus to be available */
+	while ((i2c_rcar_read(config, RCAR_I2C_ICMCR) & RCAR_I2C_ICMCR_FSDA) && (timeout < 10)) {
+		k_busy_wait(USEC_PER_MSEC);
+		timeout++;
+	}
+	if (timeout == 10) {
+		return -EIO;
+	}
+
+	do {
+		/* We are not supporting 10-bit addressing */
+		if ((msgs->flags & I2C_MSG_ADDR_10_BITS) == I2C_MSG_ADDR_10_BITS) {
+			return -ENOTSUP;
+		}
+
+		/* Send slave address */
+		if (i2c_rcar_set_addr(dev, addr, !!(msgs->flags & I2C_MSG_READ))) {
+			return -EIO; /* No ACK received */
+		}
+
+		/* Transfer data */
+		if (msgs->len) {
+			ret = i2c_rcar_transfer_msg(dev, msgs);
+			if (ret != 0) {
+				return ret;
+			}
+		}
+
```

Comment:
Hi @aaillet,
Just a question, when using tools as i2c_scan, we send a frame with no data so msgs-len is equal to 0 there, does we are sending the stop condition at this point, since you are not calling i2c_rcar_finish ?

Reply:
No we are not, i'll send a fix soon, thx